comms/src/connection_manager/dialer.rs

```diff
@@ -20,17 +20,378 @@
+        peer_connection::create_peer_connection,
+        utils::short_str,
+    },
+    multiaddr::Multiaddr,
+    peer_manager::{Peer, PeerId},
+    transports::Transport,
+    types::CommsPublicKey,
+};
+use futures::{
+    channel::{mpsc, oneshot},
+    future,
+    future::{BoxFuture, Either},
+    stream::{Fuse, FuturesUnordered},
+    AsyncRead,
+    AsyncWrite,
+    FutureExt,
+    SinkExt,
+    StreamExt,
+};
+use log::*;
+use std::{collections::HashMap, sync::Arc, time::Instant};
+use tari_shutdown::{Shutdown, ShutdownSignal};
+use tokio::{runtime::TaskExecutor, timer};
 
-/// Encapsulates the ability to asynchronously establish connections.
-pub trait Dialer<T> {
-    /// What this dialer produces when it's work is done. Typically a PeerConnection / TcpStream
-    type Output;
-    /// The error type that this dialer can produce
-    type Error;
-    /// The future returned from the dial function
-    type Future: Future<Output = Result<Self::Output, Self::Error>>;
+const LOG_TARGET: &str = "comms::connection_manager::establisher";
 
-    /// Asynchronously establish a connection (dial) to a remote peer
-    fn dial(&self, peer: &T) -> Self::Future;
+type DialResult<T> = Result<T, ConnectionManagerError>;
+
+#[derive(Debug)]
+pub enum DialerRequest {
+    Dial(Box<(Peer, oneshot::Sender<Result<PeerConnection, ConnectionManagerError>>)>),
+    CancelPending(PeerId),
+}
+
+pub struct Dialer<TTransport, TBackoff> {
+    executor: TaskExecutor,
+    config: ConnectionManagerConfig,
+    transport: TTransport,
+    backoff: Arc<TBackoff>,
+    request_rx: Fuse<mpsc::Receiver<DialerRequest>>,
+    cancel_signals: HashMap<PeerId, Shutdown>,
+    conn_man_notifier: mpsc::Sender<ConnectionManagerEvent>,
+    shutdown: Option<ShutdownSignal>,
+    pending_dial_requests: HashMap<PeerId, Vec<oneshot::Sender<Result<PeerConnection, ConnectionManagerError>>>>,
+}
+
+impl<TTransport, TSocket, TBackoff> Dialer<TTransport, TBackoff>
+where
+    TTransport: Transport<Output = (TSocket, CommsPublicKey, Multiaddr)> + Unpin + Send + Sync + Clone + 'static,
+    TSocket: AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static,
+    TBackoff: Backoff + Send + Sync + 'static,
+{
+    pub fn new(
+        executor: TaskExecutor,
+        config: ConnectionManagerConfig,
+        transport: TTransport,
+        backoff: Arc<TBackoff>,
+        request_rx: mpsc::Receiver<DialerRequest>,
+        conn_man_notifier: mpsc::Sender<ConnectionManagerEvent>,
+        shutdown: ShutdownSignal,
+    ) -> Self
+    {
+        Self {
+            executor,
+            config,
+            transport,
+            backoff,
+            request_rx: request_rx.fuse(),
+            cancel_signals: Default::default(),
+            conn_man_notifier,
+            shutdown: Some(shutdown),
+            pending_dial_requests: Default::default(),
+        }
+    }
+
+    pub async fn run(mut self) {
+        let mut pending_dials = FuturesUnordered::new();
```

Comment:
Couldn't this just be part of the Dialer struct so that you don't need to pass it around to the handlers and just call it from `self`?

Reply:
Perhaps - I had troubles with the borrow checker doing it that way in the past, but let me try again

Reply:
I did it in the Transaction Service but might be a subtle difference in the applications

Reply:
So the real reason I didn't do this is because:
1. The box future would need to be `Sync` (because connection manager is spawned) and that's an extra trait bound not needed when it's contained in the method. 
2. the type of future is `BoxFuture<'static, Option<(DialState, DialResult<TTransport::Output>)>>` - this means that the trait bound `TTransport: Transport` needs to be on the Dialer and ConnectionManager struct declarations

These are not insurmountable issues but I think that it ends up being neater to contain it within the method

Reply:
Cool cool